Improve `vec_compare()`

DavisVaughan opened the pull request:
Closes #648 

Still working on this, but incredibly promising results with a rewrite of `vec_compare()`. 

It follows the ideas of `vec_equal()` for the most part. But importantly I'm also experimenting with the idea of doing the comparisons column wise on data frames. This is way more efficient than row wise and I think we might be able to port this idea over to `vec_equal()` too.

The way it generally works is that we construct a logical `skip` vector initialized to all false values and with size equal to the number of rows in the data frame. As we go through the data frame columns, we flip the `skip` values to `TRUE` when we "know" the comparison value (i.e. if it is less than or greater than) and we don't ever evaluate that row again when comparing downstream columns. If all `skip` values are ever `TRUE` (as measured by a decreasing `count` that hits `0` when we know everything), we exit immediately.

Before

``` r
library(vctrs)
library(nycflights13)

bench::mark(
  vec_compare(flights$year, flights$year)
)
#> # A tibble: 1 x 6
#>   expression                                 min median `itr/sec` mem_alloc
#>   <bch:expr>                              <bch:> <bch:>     <dbl> <bch:byt>
#> 1 vec_compare(flights$year, flights$year) 5.26ms 6.48ms      156.     135MB
#> # … with 1 more variable: `gc/sec` <dbl>

bench::mark(
  vec_compare(flights, flights)
)
#> # A tibble: 1 x 6
#>   expression                        min median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                    <bch:t> <bch:>     <dbl> <bch:byt>    <dbl>
#> 1 vec_compare(flights, flights)   189ms  197ms      5.07    58.2MB     2.54

flights_sort <- dplyr::arrange(flights, dep_time)

bench::mark(
  vec_compare(flights, flights_sort)
)
#> # A tibble: 1 x 6
#>   expression                            min median `itr/sec` mem_alloc
#>   <bch:expr>                         <bch:> <bch:>     <dbl> <bch:byt>
#> 1 vec_compare(flights, flights_sort) 39.1ms 42.3ms      23.9    11.6MB
#> # … with 1 more variable: `gc/sec` <dbl>
```

After

``` r
library(vctrs)
library(nycflights13)

# vector speed up
bench::mark(
  vec_compare(flights$year, flights$year)
)
#> # A tibble: 1 x 6
#>   expression                                min median `itr/sec` mem_alloc
#>   <bch:expr>                              <bch> <bch:>     <dbl> <bch:byt>
#> 1 vec_compare(flights$year, flights$year) 351µs  509µs     1700.     133MB
#> # … with 1 more variable: `gc/sec` <dbl>

# all equal "worst case"
bench::mark(
  vec_compare(flights, flights)
)
#> # A tibble: 1 x 6
#>   expression                        min median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                    <bch:t> <bch:>     <dbl> <bch:byt>    <dbl>
#> 1 vec_compare(flights, flights)  12.3ms   13ms      75.6    2.92MB     2.04

flights_sort <- dplyr::arrange(flights, dep_time)

# not all equal, we compare column wise and skip if we already
# know the result
bench::mark(
  vec_compare(flights, flights_sort)
)
#> # A tibble: 1 x 6
#>   expression                            min median `itr/sec` mem_alloc
#>   <bch:expr>                         <bch:> <bch:>     <dbl> <bch:byt>
#> 1 vec_compare(flights, flights_sort) 2.52ms 3.13ms      306.    2.57MB
#> # … with 1 more variable: `gc/sec` <dbl>
```

<sup>Created on 2019-11-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

lionel- commented:
> I'm also experimenting with the idea of doing the comparisons column wise on data frames. This is way more efficient than row wise and I think we might be able to port this idea over to vec_equal() 

yup we definitely need to finish vectorising data frame operations in vctrs, would be great to reach this point!

DavisVaughan commented:
@lionel- ready for review

- Added encoding support for `vec_compare()`, which I had forgotten about when I did it for `vec_equal()`
- We validate `na_equal` now
- Added various tests to flesh out existing coverage

There is a big drop in test coverage for this file because `compare_scalar()` is no longer used by anything. This means it has no tests, along with `df_compare_scalar()`, which is only ever used by `compare_scalar()`. We include `compare_scalar()` in the vctrs.h api file (like `equal_scalar()`), but we don't actually use it.

Not sure what to do about that. It seems like it might someday be useful internally, but I don't really feel like it would be worth it to export it to the R side just for testing now.

lionel- commented on src/compare.c:
```diff
@@ -1,136 +1,325 @@
+static SEXP df_compare_impl(SEXP x, SEXP y, bool na_equal, SEXP info, R_len_t n);
+
+static SEXP df_compare(SEXP x, SEXP y, bool na_equal, R_len_t n) {
+  // Skip nothing to begin with
```
It would help to explain the algorithm here, e.g. what `skip` is for and what `count` is counting.

Maybe `count` should be `remaining_rows` and `skip` something like `row_done` to indicate it's about rows that are already compared?

lionel- commented on src/compare.c:
```diff
@@ -1,136 +1,325 @@
+static SEXP vec_compare_col(SEXP x, SEXP y, bool na_equal, SEXP info, R_len_t n);
+static SEXP df_compare_impl(SEXP x, SEXP y, bool na_equal, SEXP info, R_len_t n);
+
+static SEXP df_compare(SEXP x, SEXP y, bool na_equal, R_len_t n) {
```
Maybe rename `n` to `nrow` everywhere, to help keep track of what it's about?

lionel- commented on tests/testthat/test-compare.R:
```diff
@@ -107,6 +147,59 @@ test_that("vec_proxy_compare() handles data frame with a POSIXlt column", {
+})
+
+test_that("equality can always be determined when strings have identical encodings", {
+  encs <- c(encodings(), list(bytes = encoding_bytes()))
```
I was puzzled for a bit here. I wonder if we should switch to this clearer way of combining lists:

```
encs <- list2(!!!encodings(), bytes = encoding_bytes())
```

lionel- commented on tests/testthat/test-compare.R:
```diff
@@ -107,6 +147,59 @@ test_that("vec_proxy_compare() handles data frame with a POSIXlt column", {
+  }
+})
+
+test_that("can compare different strings with different encodings", {
```
```suggestion
test_that("can compare non-equal strings with different encodings", {
```

lionel- commented on tests/testthat/test-compare.R:
```diff
@@ -55,11 +55,51 @@ test_that("data frames are compared column by column", {
+  # Names are not checked, as `vec_cast_common()` should take care of the type.
+  # So if `vec_cast_common()` is not called, or is improperly specified, then
+  # this could result in false equality.
+  expect_equal(.Call(vctrs_compare, df, setNames(df, c("x", "z")), TRUE), c(0, 0, 0))
```
Maybe also a test case for unequal rows?

```r
  df1 <- new_data_frame(list(x = 1:3, y = 1))
  df2 <- new_data_frame(list(y = 1:2, x = 1:2))
  .Call(vctrs_compare, df1, df2, TRUE)
```

lionel- commented on src/compare.c:
```diff
@@ -1,136 +1,325 @@
-        }
-      }
+  switch (vec_proxy_typeof(x)) {
+  case vctrs_type_list: stop_not_comparable(x, y, "lists are not comparable");
```
It seems weird that we can compare list-columns but not lists, like an invariant is broken. Perhaps we should be consistent.

DavisVaughan replied:
I was _very_ surprised that we could do a `vec_compare()` on a data frame with list columns. This test was meant to be one saying "cannot compare data frames with list columns" until I realized it worked. 

https://github.com/r-lib/vctrs/pull/649/files#diff-190e0bcf1adfd9d01b92248770e25c0cR85

We don't actually do a true comparison. The list column gets replaced with `vec_seq_along()` in both `x` and `y`, so essentially it gets skipped.

This happens in `vec_proxy_compare.data.frame()` where we set `vec_proxy_compare(relax = TRUE)`.

lionel- replied:
This makes sense when you consider `arrange()` on a data frame containing list columns. We can't have an error there, and a descent into list elements would be costly and would not produce meaningful ordering.

DavisVaughan commented on src/compare.c:
```diff
@@ -1,136 +1,325 @@
-    }
+// -----------------------------------------------------------------------------
+
+static int lgl_compare_scalar(const int* x, const int* y, bool na_equal) {
```
✔️

DavisVaughan commented on src/compare.c:
```diff
@@ -1,136 +1,325 @@
+  int* p_out = INTEGER(out);
+  memset(p_out, 0, n * sizeof(int));
+
+  SEXP info = PROTECT(Rf_allocVector(VECSXP, 3));
```
✔️

lionel- commented on src/compare.c:
```diff
@@ -216,37 +219,82 @@ SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_equal_) {
+} while (0)
+
+static struct vctrs_df_compare_info init_compare_info(R_len_t n_row) {
+  struct vctrs_df_compare_info info;
```
This seems a bit overkill for a one-off initialisation, but that's not a big deal.

lionel- commented on src/compare.c:
```diff
@@ -216,37 +219,82 @@ SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_equal_) {
+ * @member remaining The number of `row_known` values that are still `FALSE`.
+ *   If this hits `0` before we traverse the entire data frame, we can exit
+ *   immediately because all comparison values are already known.
+ */
```
👍
